### DISCUSSION TITLE
Build error "Undefined symbols" when building with internal type in a package

### DISCUSSION THREAD

**Discussion by gsl-anthonymerle** [pointfreeco/swift-composable-architecture#2667] [General]
### Description

Before anything, I want to say that I post the issue onthe TCA repository because I could not reproduce my case without using Reducers, but I do think this is more tied to some compiler bug than TCA 🙏 

In a somewhat specific scenario, building an application fails with the following error:
> ld: Undefined symbols:
  nominal type descriptor for MyTemporaryPackage.PackageFeature.SomeThing, referenced from:
      _symbolic _____y_______________y_____ytGG 22ComposableArchitecture5ScopeV 18MyTemporaryPackage0F7FeatureV5StateV AF6ActionO AA12EmptyReducerV AF9SomeThingV in SomeFeature.o
clang: error: linker command failed with exit code 1 (use -v to see invocation)

Here is the setup to reproduce:

- Having an Application (MyApp) project with a local package (MyPackage)
- Having a reducer in MyPackage
- Having a reducer in MyApp that uses the MyPackage's reducer
- Having a piece of MyPackage's reducer's State to use an **internal** type (for example an internal error that is used by the reducer to show some things)
- Having to MyPackage's reducer's body, scope the internal piece of state
- **building for a release config, not a debug one**

I've made a small project that reproduces the issue: https://github.com/gsl-anthonymerle/TCABuildError

Side notes:

- I unfortunately could not reproduce this issue without TCA. I even tried duplicating the Reducer code into a smaller one to at least not import TCA but without any success
- I suspect a compiler issue because if I change the release's build settings to not optimise the symbols it seems to work fine (SWIFT_OPTIMIZATION_LEVEL=-Onone)
- The **internal** type seems to be a key because it compiles fine if it is either: public, package, or even private

### Checklist

- [X] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

I should not have a compilation error when build for release

### Actual behavior

I have a compilation error when build for release

### Steps to reproduce

Clone this project: https://github.com/gsl-anthonymerle/TCABuildError
Build for release

### The Composable Architecture version information

1.5.3

### Destination operating system

iOS 15

### Xcode version information

Version 15.0.1 (15A507)

### Swift Compiler version information

```shell
swift-driver version: 1.87.1 Apple Swift version 5.9 (swiftlang-5.9.0.128.108 clang-1500.0.40.1)
Target: arm64-apple-macosx14.0
```

**Comment by mbrandonw** [comment]
Hi @gsl-anthonymerle, this is definitely a Swift compiler bug. You will notice that you do not get the error if you remove the `Scope` from `PackageFeature`:

```swift
public var body: some ReducerOf<Self> {
  EmptyReducer()
}
```

You can even keep the `Scope`, but remove the reference to the internal state:

```swift
public var body: some ReducerOf<Self> {
  Scope(state: \.self, action: \.self) {
    EmptyReducer()
  }
}
```

Even that compiles.

This is a bug with constrained opaque types (e.g. `some ReducerOf<Self>`) in Swift where you cannot reference an internal or private type inside the `some` type. We [filed a bug](https://github.com/apple/swift/issues/69927) with Apple about this.

I don't think there's anything TCA can do to mitigate this. Probably the reason you could not reproduce it without TCA is because you didn't recreate the full machinery of the builders and opaque types, which is necessary to actually see the bug. If you were to do that then I'm sure you would see the same problem without having to import TCA.

Since this is not an issue with the library I am going to convert it to a discussion. We also highly recommend you file this bug with Apple.

**Reply by gsl-anthonymerle** [reply]
Hi @mbrandonw,

Thanks a lot for your answer, I didn't see that you already filed a bug to Apple for the same issue 🙏 
I will link the repository I created so that they also have a smaller context than Isowords to work on 🤞 

Have a nice day,
Anthony

**Comment by nayooti** [comment]
I had this too and I got to a point where it became impossible to maintain my project. Virtually every new TCA project broke my project + tests. 
The only solution that worked in my case was to **turn all my frameworks into swift packages**.

**Reply by mbrandonw** [reply]
@nayooti Did you see the comment above about making the types public fixes the problem? You can even underscore the types so that they don't autocomplete, and you can keep the `init`s internal/private so that other modules can't create the types. That seems like the simplest solution to me.
